wasmcloud-chat/actors/broker-channel/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+const OP_PROCESS_MESSAGE: &str = "ProcessMessage";
+
+const CHANNEL_ID: &str = "messagebroker";
+const MSGTYPE_MESSAGE: &str = "MSG";
+
+const LINK_NAME_FRONTEND: &str = "frontend";
+const _LINK_NAME_BACKEND: &str = "backend";
+
+#[actor::init]
+fn init() {
+    messaging::Handlers::register_handle_message(handle_message);
+}
+
+fn handle_message(msg: messaging::BrokerMessage) -> HandlerResult<()> {
+    // TODO: handle request to obtain access token
+
+    if msg.subject == INBOUND_SUBJECT {
+        handle_inbound_message(msg)
+    } else if msg.subject.starts_with(BACKEND_SUBJECT_PREFIX) {
+        handle_outbound_message(msg)
+    } else {
+        Err(format!("Unrecognized subject {}", msg.subject).into())
+    }
+}
+
+/// Validates that the inbound request has a legitimate security authorization to
+/// publish messages.
+/// Takes an `InboundRequest` message delivered from the `frontend` link binding
+/// and converts that to a `ChannelMessage`, to be used in directly invoking
+/// the `messages` actor's `ProcessMessage` operation. Responds to the front-end
+/// with a copy of the acknowledgement received from the internal actor.
+fn handle_inbound_message(msg: messaging::BrokerMessage) -> HandlerResult<()> {
+    // TODO: validate access token
+    let req: InboundRequest = serde_json::from_slice(&msg.body)?;
+    let new_guid = extras::default()
+        .request_guid()?
+        .unwrap_or("???".to_string());
+    let channelmsg = msgactor::ChannelMessage {
+        created_on: req.created_on,
+        message_id: new_guid,
+        data: None,
+        origin_channel: CHANNEL_ID.to_string(),
+        origin_user_id: req.origin_user_id,
+        origin_room: req.origin_room,
+        target_url: req.target_url,
+        message_type: MSGTYPE_MESSAGE.to_string(),
+        message_text: req.message_text,
+    };
+    let ack: msgactor::ProcessAck =
+        actor::call_actor(MSG_ACTOR_CALL_ALIAS, OP_PROCESS_MESSAGE, &channelmsg)?;
+    let response = InboundResponse {
+        acknowledged: ack.processed,
+        error: ack.error,
+        message_id: ack.message_id,
+    };
+    if !msg.reply_to.is_empty() {
+        let _ = messaging::host(LINK_NAME_FRONTEND).publish(
+            msg.reply_to,
+            "".to_string(),
+            serde_json::to_vec(&response)?,
+        )?;
+    }
+    Ok(())
+}
+
+/// Takes a CloudEvent received from the back end and simply forwards it to the front-end
+/// by publishing it on the `frontend` link binding and replacing the subject prefix
+fn handle_outbound_message(msg: messaging::BrokerMessage) -> HandlerResult<()> {
+    let cloud_event: CloudEvent = serde_json::from_slice(&msg.body)?;
+    let subject = msg
+        .subject
+        .replace(BACKEND_SUBJECT_PREFIX, FRONTEND_SUBJECT_PREFIX);
+    let _ = messaging::host(LINK_NAME_FRONTEND).publish(
+        subject,
+        "".to_string(),
+        serde_json::to_vec(&cloud_event)?,
+    )?;
```

Comment:
So this explicit deserialize/serialize is to make sure that the data isn't just blindly forwarded, it's asserted to be a CloudEvent? That's a neat pattern.

Reply:
Exactly.

By unraveling here we also now have a good spot where we can add augmentation / decoration to the cloud event during proxying if we choose.